lib/web_ui/lib/src/engine/configuration.dart

```diff
@@ -51,14 +52,49 @@ void debugSetConfiguration(FlutterConfiguration configuration) {
 class FlutterConfiguration {
   /// Constructs a configuration from a JavaScript object containing
   /// runtime-supplied properties.
-  FlutterConfiguration(this._js);
+  FlutterConfiguration(JsFlutterConfiguration? config) {
+    addConfigurationOverrides(config);
+  }
 
-  final JsFlutterConfiguration? _js;
+  final List<JsFlutterConfiguration> _configurations = <JsFlutterConfiguration>[];
+
+  /// Adds a configuration [overrides] object that will take precedence over
+  /// any existing configuration.
+  ///
+  /// This is used so we support multiple sources of configuration with a consistent
+  /// precedence. From lowest to highest:
+  ///
+  /// * default static values
+  /// * window.flutterConfiguration object
+  /// * engineInitializer.initEngine(runtimeConfig)
+  ///
+  /// Overrides are always added at position [0] of the array, so the order in
+  /// which this method is called (first with flutterConfiguration), then with
+  /// the runtimeConfig, matters.
+  void addConfigurationOverrides(JsFlutterConfiguration? overrides) {
+    if (overrides != null) {
+      _configurations.insert(0, overrides);
+    }
+  }
+
+  T getConfigValue<T>(T? Function(JsFlutterConfiguration) extractor, T defaultValue) {
+    try {
+      final JsFlutterConfiguration config = _configurations
+          .firstWhere(
+            (JsFlutterConfiguration config) => extractor(config) != null
+          );
+      return extractor(config)!; // Guaranteed not null, because two lines above.
+    } on StateError {
```

Comment:
Avoid using throwing exceptions for normal control flow scenarios. There's nothing wrong with runtime configs missing a value and needing to default to the `defaultValue`. Therefore no exceptions should be thrown. Consider using `orElse`, or just a plain loop.

See also: https://dzone.com/articles/exceptions-as-controlflow-in-java

Reply:
This should have been a `firstWhereOrNull` or a plain loop, but now it's gone forever!